### ISSUE TITLE
[MPS] 3D `F.linear` on `[B,1,K]` (seq=1 decode) is up to 8.5x slower than 2D / `torch.matmul` (bf16/fp16)

### ISSUE THREAD

**Issue by radimurban** [pytorch/pytorch#189847]
### 🐛 Describe the bug

On Apple Silicon MPS, `torch.nn.functional.linear` applied to a 3D input of decode shape `[B, 1, K]` is up to ~8.5x slower than the mathematically identical 2D op `[B, K]`, and than `torch.matmul(x, w.t())` on the same 3D tensor. The cost ramps ~linearly with `B` and then snaps back to GEMM once `B` crosses a threshold (~16). It directly throttles on-device LLM token generation

It is specific to
 - MPS
- `F.linear` (which on MPS routes to `_mps_linear`)
- half precision (fp32 is unaffected)
- 3D half input with small non-batch dims (`M == 1` is the worst/most-persistent case)

## Environment

```
PyTorch : 2.14.0a0 (source build; also present on 2.10.x)
OS      : macOS 26.4 (25E246)
Device  : Apple M4 Pro (MPS)
dtype   : bfloat16 (fp16 identical; fp32 unaffected)
```

## Minimal reproduction

```python
import time, torch
import torch.nn.functional as F
dev, dt = "mps", torch.bfloat16
K, N = 5120, 17408                              # Qwen3-14B gate: [.,K] @ [K,N]
W = torch.randn(N, K, device=dev, dtype=dt)     # nn.Linear weight [out, in]

def timed(fn, warmup=10, iters=30):
    for _ in range(warmup): fn()
    torch.mps.synchronize(); ts=[]
    for _ in range(iters):
        torch.mps.synchronize(); t=time.perf_counter()
        fn(); torch.mps.synchronize(); ts.append((time.perf_counter()-t)*1e3)
    return sorted(ts)[len(ts)//2]

print(f"{'B':>4} {'F.linear_3d':>12} {'F.linear_2d':>12} {'matmul_3d':>10} {'3d/2d':>6}")
for B in (1,4,8,14,16,32,64):
    x3 = torch.randn(B,1,K,device=dev,dtype=dt)  # [B,1,K] decode shape
    x2 = x3.squeeze(1)                            # [B,K] (zero-copy view)
    l3 = timed(lambda: F.linear(x3, W))
    l2 = timed(lambda: F.linear(x2, W))
    m3 = timed(lambda: torch.matmul(x3, W.t()))
    print(f"{B:>4} {l3:>12.2f} {l2:>12.2f} {m3:>10.2f} {l3/l2:>5.1f}x")
```

Output (M4 Pro, macOS 26.4, bf16):

```
   B  F.linear_3d  F.linear_2d   matmul_3d  3d/2d
   1         0.89         0.91        0.93   1.0x
   4         3.00         0.97        0.97   3.1x
   8         5.76         0.94        0.93   6.1x
  14         9.91         1.17        1.18   8.5x
  16         1.15         1.14        1.15   1.0x
  32         1.28         1.29        1.30   1.0x
  64         1.97         1.95        1.96   1.0x
```

- **Expected:** `F.linear([B,1,K])` costs ~the same as `F.linear([B,K])` and
  `matmul([B,1,K], w.t())` — a `[B,1]×N` output is a trivial reshape from a
  `[B]×N` GEMM.
- **Actual:** `F.linear([B,1,K])` ramps ~linearly in `B` (each batch element run
  as an independent GEMV) and only snaps to GEMM performance past a width-
  dependent `B` threshold. `matmul` and 2D `F.linear` stay flat.

## Analysis

<img width="3552" height="2648" alt="Image" src="https://github.com/user-attachments/assets/0c1f7825-043e-446b-b400-c0245b2b1d25" />

1. **`F.linear` only.** `torch.matmul`, `torch.mm`, `torch.addmm` on the same logical op do **not** show the linear-per-batch ramp.
2. **3D half input with small non-batch dims.** `M == 1` (`[B,1,K]`) is the worst and most persistent case, but `[B,2,K]` also ramps at low `B` and snaps earlier (~B=5); `[B,8,K]` tracks the efficient GEMM sooner. A leading singleton `[1,B,K]` (middle dim `= B`) is flat. The pathology is the small-leading-dims 3D layout staying below the `N`-dependent GEMM threshold; seq=1 decode (`M==1`) sits in it for the widest `B` range.
3. **Half precision only.** At `B=8` gate: bf16 = 6.1x, fp16 = 6.0x, fp32 = 1.0x (no cliff).
4. **Threshold scales with output width `N`.** Snap-back `B` is `None` only for  `N = 512`; ~16 for `N ∈ [1024, 17408]`; ~10 for `N ≥ 32768`. So the per-layer snap differs by width (MLP `N=17408` ~16, `lm_head` `N=151936` ~10), and a full decode step's summed
   cost peaks around batch 8–14 before the widest ops flip to GEMM.

## Source localization

`aten::linear` on MPS maps to `_mps_linear` (`native_functions.yaml` → `aten/src/ATen/native/mps/operations/Linear.mm`)

For the decode case (macOS 15+, contiguous, non-complex, half) `_mps_linear` takes the **nograph fast path** `_mps_linear_nograph`, which hands the raw 3D `[B,1,K]` NDArray to `MPSNDArrayMatrixMultiplication` (weight transposed, broadcast over the leading batch dim) → per-batch GEMV.

A flatten-to-2D perf workaround for exactly this 3D case already exists, but only in the graph branch of `_mps_linear` (reached on macOS < 15):

```cpp
// graph branch -- the perf workaround exists here:
bool doReshape = input.dim() > 4;
if (!doReshape && is_bias_defined) {
  // workaround to improve the performance with 3D+ inputs
  doReshape = input_size.size() > 2 && input_size[0] > 1 &&
              input_size[1] >= 1 && input_size[1] <= 32 && bias.dim() <= 1;
}
```

The `nograph` fast path has no equivalent perf flatten. Its only flatten, `needs_nd_workaround(input)`, is a non-determinism fix gated to the Apple10 (M5+) GPU family (#180776) — it does not fire on M4 and is unrelated to this GEMV perf issue. So on M4 / macOS 26 a contiguous half `[B,1,K]` reaches `_mps_linear_nograph` with no flatten → the per-batch GEMV cliff.

## Proposed fix

The `nograph` fast path already has a flatten-to-2D branch, it is just gated to the M5+ non-determinism case (`needs_nd_workaround`). Broadening that gate to any `>2D` input fixes the GEMV cliff and subsumes the existing workaround. One-line change in `_mps_linear`:

```diff
-    if (needs_nd_workaround(input) && (!kernel_bias.defined() || kernel_bias.dim() <= 1)) {
+    if (input.dim() > 2 && (!kernel_bias.defined() || kernel_bias.dim() <= 1)) {
       auto input2d = input.flatten(0, -2);    // [B,1,K] -> [B,K]  (zero-copy view)
       auto output2d = output.flatten(0, -2);  // [B,1,N] -> [B,N]
       _mps_linear_nograph(input2d, weight, kernel_bias, output2d);
     } else {
       _mps_linear_nograph(input, weight, kernel_bias, output);
     }
```

`flatten(0, -2)` on a contiguous input is a zero-copy view.

## Verification

**(a) Related PR #184631 (drop `strides()` in `_mps_linear_nograph`) does not fix this.** A canonically-strided `[B,1,K]` never takes the strided NDArray path that PR targets, so the cliff is unchanged: `F.linear_3d` still 6.2x at B=8, 8.5x at B=14, snapping at 16. (That PR fixes a *different* problem — non-canonically strided contiguous inputs, e.g. a B=1 transpose in `MultiheadAttention`, #181725.)

**(b) The flatten fix above eliminates it.** With the gate broadened, `F.linear` on `[B,1,K]` matches 2D and `torch.matmul` at every batch, with no large-batch regression (M4 Pro, bf16, ms):

```
   B | F.linear_3d | F.linear_2d | matmul_3d | 3d/2d
   1 |        0.94 |        0.94 |      0.97 | 1.0x
   8 |        0.93 |        0.93 |      0.92 | 1.0x    (was 5.8 ms / 6.2x)
  14 |        1.16 |        1.16 |      1.18 | 1.0x    (was 9.9 ms / 8.5x)
  16 |        1.17 |        1.17 |      1.18 | 1.0x
  32 |        1.29 |        1.29 |      1.31 | 1.0x
  64 |        1.97 |        1.96 |      1.97 | 1.0x
```

Output is bit-identical (`torch.allclose` / `torch.equal`), since the flatten is a pure reshape.

## Related Issues

- #186541 (open, "Port all MPSGraph matmuls to Metal") — benchmark table shows MPSGraph matmul slow on GEMV (M==1) / small-batch bmm.
- #184631 (open, "Skip strided NDArray path in `_mps_linear_nograph` when inputs are contiguous") — same function, but a *different* cause (non-canonical
  strides). **Tested: does not fix this GEMV cliff** (see Verification).
- #181725 (open) — `nn.MultiheadAttention` ~9x slower at B=1; the case #184631 fixes (B=1 strided `F.linear`), distinct from this M=1 GEMV cliff.
- #188728 (open) — `F.linear` ignores `PYTORCH_MPS_PREFER_METAL`.
- #117837 (merged) — origin of the graph-branch `input_size[1] <= 32` flatten workaround. #180776 (closed) / #181466 (merged) — the M5+ >2D half `linear` non-determinism flatten this fix generalizes.

### Versions

PyTorch version: 2.14.0a0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.4.2 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.1.1.101)
CMake version: version 4.2.3
Libc version: N/A

Python version: 3.12.13 | packaged by Anaconda, Inc. | (main, Mar 19 2026, 20:12:32) [Clang 20.1.8 ] (64-bit runtime)
Python platform: macOS-26.4.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Apple M3 Max

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] torch==2.14.0a0
[conda] numpy                      2.4.4                pypi_0           pypi
[conda] torch                      2.14.0a0  pypi_0           pypi

---

CC: @jhavukainen , @jerryzh168 , @versi379 , @kulinseth , @malfet , @DenisVieriu97 , @aditvenk

**Comment by fenilsonani** [comment]
Independently hit this while benchmarking LLM decode (nvidia/canary-qwen-2.5b, a Qwen3-1.7B decoder) on MPS — confirming on different hardware and a stock release build, plus measured end-to-end impact.

**Environment:** Apple M-series (36 GB), macOS 26 (Darwin 25.5.0), stock **torch 2.13.0** from PyPI, Python 3.11.15 — so this is in current release wheels, not just source/nightly builds.

**Microbenchmark** (`F.linear`, B=8, 3D `[B,1,K]` vs 2D `[B,K]`, mean over 30 iters after warmup, Qwen3-1.7B layer shapes):

| shape | dtype | 2D ms | 3D ms | 3D/2D |
|---|---|---|---|---|
| mlp up 2048→6144 | fp32 | 0.308 | 0.224 | 0.73x |
| mlp up 2048→6144 | fp16 | 0.067 | 0.283 | **4.24x** |
| mlp up 2048→6144 | bf16 | 0.067 | 0.284 | **4.22x** |
| mlp down 6144→2048 | bf16 | 0.065 | 0.287 | **4.44x** |
| lm_head 2048→151936 | fp32 | 4.711 | 4.710 | 1.00x |
| lm_head 2048→151936 | fp16 | 1.635 | 13.422 | **8.21x** |
| lm_head 2048→151936 | bf16 | 1.710 | 13.200 | **7.72x** |

Matches the reported pattern: half precision only, B=1 unaffected, and the penalty is gone by B=32 — consistent with the per-batch-GEMV ramp and snap-back threshold described above.

**End-to-end impact** (HF `generate`, greedy, KV cache, 64 new tokens, batch 8, Qwen3-1.7B): bf16 decodes at **68 ms/tok** stock vs **44 ms/tok in fp32** — i.e. the user-visible symptom is "half precision is *slower* than fp32 at batch 2–8, faster at 16+", which is how we noticed. With a reshape-to-2D patch on `nn.Linear.forward`:

```python
def linear_2d_mps(self, x):
    if x.dim() == 3 and x.device.type == "mps" and x.dtype in (torch.float16, torch.bfloat16):
        b, s, k = x.shape
        return F.linear(x.reshape(b * s, k), self.weight, self.bias).reshape(b, s, -1)
    return _orig_forward(self, x)
```

bf16 drops to **26 ms/tok** (2.6x vs stock bf16, 1.7x vs fp32), with bit-identical outputs. Since `[B,1,K]` is exactly what HF `generate` feeds every linear layer per decode step, effectively every half-precision LLM decoding on MPS at batch 2–15 is paying this.

Happy to test candidate fixes on this workload.

**Comment by radimurban** [comment]
Hey @fenilsonani, thanks for the comment. You can test #189855 - it should remove that regression.
